### PR TITLE
Implement and test top-level all_of

### PR DESCRIPTION
**Pull request by ayousufi**
This is a follow up to the schema change in https://github.com/block/elasticgraph/pull/355. While this PR mostly updates the unit, integration, and acceptance tests, there is a key implementation change in the filter interpreter such that `all_of` filters will now build a new bool sub-filter for each subexpression and push it onto the parent’s `:filter` array. This ensures an ANDing across the subexpressions.

**Performance benchmarking**
Given the above change will result in an extra bool filter term, I performed benchmarking using the `test_es_query.rb` script adapted to an all_of query with nested any_satisfy blocks.

Results from current schema:
Average took value: 504ms
Median took value: 454ms
P99 took value: 1190ms

vs PR schema:
Average took value: 505ms
Median took value: 448ms
P99 took value: 1054ms

We can see there is no significant impact to performance with the additional bool filter terms.

### CODE REVIEW

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 1]
```diff
@@ -973,6 +973,120 @@ def expect_error_from(filter, *error_snippets)
...
+            },
+            "options" => {
+              "color" => {
+                "equal_to_any_of" => [enum_value(:BLUE)]
+              }
+            }
+          }
+
+          side_by_side_results = list_widgets_with(:name, filter: filter_side_by_side, order_by: [:name_ASC])
+          expect(side_by_side_results).to match([
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/acceptance/search_spec.rb:1010]: `match` performs "fuzzy" matching--it allows matchers/regexes/etc to be used within the expected datastructure rather than using string equality. Here it seems like strict equality should work, though--can you use `eq` instead?

(And same with other expectations you've added.)

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 2]
```diff
@@ -973,6 +973,120 @@ def expect_error_from(filter, *error_snippets)
...
+          w1 = build(:widget, id: "widg1", tags: ["alpha", "beta"])
+          w2 = build(:widget, id: "widg2", tags: ["alpha"])
+          w3 = build(:widget, id: "widg3", tags: ["gamma", "beta"])
+          index_records(w1, w2, w3)
+
+          filter_all_of_tags = {
+            "tags" => {
+              "all_of" => [
+                {"any_satisfy" => {"equal_to_any_of" => ["alpha"]}},
+                {"not" => {"any_satisfy" => {"equal_to_any_of" => ["beta"]}}}
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/acceptance/search_spec.rb:1047]: I believe this can be done w/o `all_of`:

```ruby
"tags" => {
  "any_satisfy" => {"equal_to_any_of" => ["alpha"]},
  "not" => {"any_satisfy" => {"equal_to_any_of" => ["beta"]}}
}
```

Does this test really exercise any new code paths or edge cases?  If not, we can maybe drop it; the acceptance tests aren't meant to be exhaustive in the same way as unit or integration tests. We do want every query feature covered, but we don't need multiple permutations of the same functionality covered in the acceptance tests.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 3]
```diff
@@ -296,6 +296,98 @@ def search_datastore(**options, &before_msearch)
           end
         end
 
+        context "top-level usage on distinct fields" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:299]: What's notable about this being `top-level usage`?  Wouldn't the behavior described below apply anywhere `all_of` is used?

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 4]
```diff
@@ -296,6 +296,98 @@ def search_datastore(**options, &before_msearch)
...
+                }
+              ]
+            }
+
+            results_all_of = search_with_freeform_filter(filter_all_of)
+            expect(results_all_of).to match_array(ids_of(t1))
+          end
+        end
+
+        context "top-level usage on the same key" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:336]: Here too I don't understand why this only apples to top-level usage.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 5]
```diff
@@ -296,6 +296,98 @@ def search_datastore(**options, &before_msearch)
...
+            )
+
+            all_of_filter = {
+              "forbes_valuations" => {
+                "all_of" => [
+                  {"any_satisfy" => {"gt" => 2_000_000}},
+                  {"any_satisfy" => {"lt" => 5_000_000}}
+                ]
+              }
+            }
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:352]: This looks like the kind of case `all_of` was already supporting. I suspect this may duplicate some existing test coverage--can you check?

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 6]
```diff
@@ -296,6 +296,98 @@ def search_datastore(**options, &before_msearch)
...
+
+            results = search_with_freeform_filter(all_of_filter)
+            expect(results).to match_array(ids_of(t1, t3))
+          end
+        end
+
+        context "combining multiple `any_of` subfilters" do
+          it "handles AND of OR blocks (allOf of anyOf)" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:360]:
```suggestion
          it "handles AND of OR's subfilters (allOf of anyOf)" do
```

(We don't use the term "blocks" to refer to subfilters anywhere else...)

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 7]
```diff
@@ -296,6 +296,98 @@ def search_datastore(**options, &before_msearch)
...
+        end
+
+        context "top-level usage on the same key" do
+          it "allows ANDing two subfilters on the same field" do
+            index_into(
+              graphql,
+              t1 = build(:team, id: "t1", forbes_valuations: [1_000_000, 100_000_000]),
+              build(:team, id: "t2", forbes_valuations: []),
+              t3 = build(:team, id: "t3", forbes_valuations: [3_000_000])
+            )
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:343]: If we keep this test (a question I put in another comment), we could make this more robust by adding a team with `forbes_valuations: [1_000_000]` and anther team with `forbes_valuations: [100_000_000]` as that would demonstrate that `all_of` is in fact treated as an `AND`. If it was treated as an `OR` the test could still pass as written.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 8]
```diff
@@ -364,6 +364,116 @@ class GraphQL
       end
 
       describe "`all_of` operator" do
+        context "top-level usage on distinct fields" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb:367]: (same question about what's special about "top-level usage" in the specs in this file...)

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 9]
```diff
@@ -364,6 +364,116 @@ class GraphQL
...
+          it "behaves the same as specifying multiple fields in the same object" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb:368]: It's confusing that this test description says it behaves the same as another case...but then doesn't demonstrate 2 cases behaving the same.

But at the level of "what payload is produced?" (the behavior we're testing with these unit tests), it doesn't behave the same. Would be good to describe this more accurately. Maybe this?

```ruby
it "produces a separate bool sub-filter for each clause"
```

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 10]
```diff
@@ -364,6 +364,116 @@ class GraphQL
...
+        end
+
+        context "top-level usage on the same key" do
+          it "allows ANDing two subfilters on the same field (otherwise impossible in a single object)" do
+            query = new_query(filter: {
+              "all_of" => [
+                {"color" => {"equal_to_any_of" => ["RED"]}},
+                {"color" => {"not" => {"equal_to_any_of" => ["GREEN"]}}}
+              ]
+            })
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb:402]: I believe this is possible w/o `all_of`:

```ruby
filter: {
  "color" => {
    "equal_to_any_of" => ["RED"],
    "not" => {"equal_to_any_of" => ["GREEN"]}
  }
}
```

...so it's a bit confusing to read it's otherwise impossible in your test description.

Maybe just omit that part of the test description?  While users don't need to use `all_of` to achieve this filtering behavior, they can use it, and you're testing what happens when they do.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 11]
```diff
@@ -364,6 +364,116 @@ class GraphQL
...
+                    }
+                  }
+                ]
+              }
+            )
+          end
+        end
+
+        context "combining multiple any_of subfilters" do
+          it "allows an AND of multiple OR blocks (i.e. allOf of anyOfs)" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb:434]: The `it` and `context` seem to be saying the same thing. Usually `context` is used to group multiple examples that share some characteristic or need shared setup or helper methods. In this case that doesn't apply. This could just be:

```ruby
it "allows an AND of multiple OR'd subfilters (i.e. allOf of anyOfs)" do
end
```

...with no `context`.

**Review by myronmarston — CHANGES_REQUESTED** [review]
Looking good! Left some feedback.

**Review by myronmarston — COMMENTED** [review]
Also, can we add some coverage for how `all_of` interacts with `not` and empty filters?

File: elasticgraph-graphql/spec/acceptance/search_spec.rb
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb:1]: I moved the all_of tests into the above test.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb (hunk 1 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Changed to `eq`.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb (hunk 2 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Removed.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 3 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Yeah not exclusive to top-level, I removed the context blocks.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 4 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Removed.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 6 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Updated.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 5 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Yeah this was a dupe, removed.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 7 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Removed per other comment.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 8 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Removed context block here as well.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 9 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Updated test name.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 11 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Removed context block.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 12]
```diff
@@ -150,6 +150,138 @@ class GraphQL
...
+              {},
+              {"forbes_valuations" => {"any_satisfy" => {"gt" => 100_000}}}
+            ]
+          }
+
+          results = search_with_freeform_filter(filter)
+          expect(results).to match_array(ids_of(t1))
+        end
+
+        it "handles empty predicates" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:255]: I think this test and the prior one can be merged into one test--index the data once, and then show two examples of empty filters.  I think of both `{}` and `field: {}` as being equivalent empty predicates--they are just two slightly different examples of the same thing, and as such these two tests aren't showing different facets of behavior.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 13]
```diff
@@ -150,6 +150,138 @@ class GraphQL
...
+        it "handles not with empty predicate" do
+          filter = {
+            "all_of" => [
+              {"not" => {}},
+              {"forbes_valuations" => {"any_satisfy" => {"gt" => 100_000}}}
+            ]
+          }
+
+          results = search_with_freeform_filter(filter)
+          expect(results).to eq([])
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:282]: The fact that no results are returned could be attributed to the fact that this test doesn't index any documents. In other words, this expectation would pass no matter what the `filter` is!

To actually demonstrate that `not: {}` within `all_of` causes the search to return no results, we need to index some documents so that there are some results that could otherwise be returned.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 14]
```diff
@@ -372,10 +476,98 @@ class GraphQL
...
-          ]}})
+          expect(datastore_body_of(query)).to query_datastore_with({bool: {
+            filter: [
+              {bool: {filter: [{terms: {"tags" => ["a", "b"]}}]}},
+              {bool: {filter: [{terms: {"tags" => ["c", "d"]}}]}}
+            ]
+          }})
+        end
+
+        it "handles not operator" do
```
Comment (myronmarston) [inline elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb:487]:
```suggestion
        it "handles `not` within `all_of` operator" do
```

...and then add a test covering `not: {all_of: ... }` please :).

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 15]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
+              },
+              {
+                "any_of" => [
+                  {"tags" => {"any_satisfy" => {"equal_to_any_of" => ["def"]}}},
+                  {"amount_cents" => {"lt" => 150}}
+                ]
+              }
+            ]
+          },
+          order_by: [:amount_cents_ASC])
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Done.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 16]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
+
+        expect(widgets).to eq([
+          string_hash_of(widget1, :id, :amount_cents)
+        ])
+
+        # Test all_of with not
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb:516]: Added.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 17]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
+              {"amount_cents" => {"gt" => 50}}
+            ]
+          },
+          order_by: [:amount_cents_ASC])
+
+        expect(widgets).to eq([
+          string_hash_of(widget1, :id, :amount_cents),
+          string_hash_of(widget2, :id, :amount_cents),
+          string_hash_of(widget3, :id, :amount_cents)
+        ])
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Done.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 18]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Done.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 19]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
+
+        # Test all_of with empty predicates
+        widgets = list_widgets_with(:amount_cents,
+          filter: {
+            "all_of" => [
+              {"tags" => {}},
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Done.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb [hunk 20]
```diff
@@ -489,6 +489,89 @@ module ElasticGraph
...
+          string_hash_of(widget1, :id, :amount_cents),
+          string_hash_of(widget2, :id, :amount_cents),
+          string_hash_of(widget3, :id, :amount_cents)
+        ])
+
+        # Test all_of with not and empty predicates
+        widgets = list_widgets_with(:amount_cents,
+          filter: {
+            "all_of" => [
+              {"not" => {}},
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb]: Done.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb [hunk 21]
```diff
@@ -150,6 +150,138 @@ class GraphQL
...
+                ]
+              }
+            ]
+          }
+
+          results = search_with_freeform_filter(filter)
+          expect(results).to match_array(ids_of(t2, t3))
+        end
+
+        it "handles not operator" do
```
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Done + added new test.

File: elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 13 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/integration/elastic_graph/graphql/datastore_query/filtering_spec.rb:281]: Done.

File: elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb (hunk 14 above)
Comment (ayousufi) [reply elasticgraph-graphql/spec/unit/elastic_graph/graphql/datastore_query/filtering_spec.rb]: Done.

File: elasticgraph-graphql/spec/acceptance/search_spec.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (myronmarston) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb:516]: I see you added coverage for `not: {all_of: [...]}` to the list-filtering behavior tests below, which is great...but can we also get a test added here? `all_of` on a list field was already supported before your changes, so `not: {all_of: [...]}` there is not newly supported. Here it is newly supported and it would be nice to include it in this section for completeness.

(Not a merge blocker, could be done in a follow up PR).

**Review by myronmarston — APPROVED** [review]
👍

File: elasticgraph-graphql/spec/acceptance/search_spec.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (ayousufi) [reply elasticgraph-graphql/spec/acceptance/search_spec.rb:516]: Added.
